### PR TITLE
Gentle and progressive fp-ts migration

### PR DESCRIPTION
**Pull request by francescopersico**
## Short description
I want to collect your opinion about a different way of migrate to fp-ts v2 in a gentle and progressive way.
Don't focus on the quality or the correctness of the migration but only on the wide scenario.
**I am going to comment my own PR to better highlight important points**

## List of changes proposed in this pull request
- Gentle and progressive migration to fp-ts 2. Progressive means that fp-ts v1 and v2 coexist in the codebase until fully ported

## How to test
Describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -103,11 +103,13 @@
     "detox": "^19.7.0",
     "fbjs": "^3.0.2",
     "fp-ts": "1.17.4",
+    "fp-ts2": "npm:fp-ts@2.12.1",
     "front-matter": "^4.0.2",
     "hastscript": "^7.0.2",
     "hoist-non-react-statics": "^3.0.1",
     "io-react-native-zendesk": "https://github.com/pagopa/io-react-native-zendesk.git#0.3.21",
     "io-ts": "1.8.5",
+    "io-ts2": "npm:io-ts@2.2.16",
```
Comment (francescopersico) [inline package.json:112]: Same as above.

File: scripts/ts/checkOutdatedDependencies/checkOutdatedDependencies.ts [hunk 2]
```diff
@@ -12,6 +12,9 @@ import { DependenciesTable } from "./types/DependeciesTable";
 import { getSeverityType } from "./types/GroupBySeverity";
 import { getDependencyType, increaseSeverityAmount } from "./types/GroupByType";
 import { OutdatedStats, toOutdatedPackage } from "./types/OutdatedStats";
+import { pipe } from "fp-ts2/lib/function";
+
+import { eitherToV2 } from "../../../ts/migration/fp-ts-converters";
```
Comment (francescopersico) [inline scripts/ts/checkOutdatedDependencies/checkOutdatedDependencies.ts:17]: When needed we can use converters v1 <-> v2.

You can remove the conversion after the code you depend on is migrated to v2. Better explained below.

File: ts/sagas/messages/loadMessage.ts [hunk 3]
```diff
@@ -11,6 +11,9 @@ import { readablePrivacyReport } from "../../utils/reporters";
 import { isTestEnv } from "../../utils/environment";
 import { convertUnknownToError } from "../../utils/errors";
 
+import * as E2 from "fp-ts2/Either";
```
Comment (francescopersico) [inline ts/sagas/messages/loadMessage.ts:14]: Another conversion example in a saga.

File: ts/sagas/messages/loadMessage.ts [hunk 4]
```diff
@@ -29,24 +32,25 @@ export function* loadMessage(
...
-    const maybeMessage: SagaCallReturnType<typeof fetchMessage> = yield* call(
+    const maybeMessageV1: SagaCallReturnType<typeof fetchMessage> = yield* call(
       fetchMessage,
       getMessage,
       meta
     );
 
-    if (maybeMessage.isLeft()) {
-      throw maybeMessage.value;
+    const maybeMessage = eitherToV2(maybeMessageV1);
```
Comment (francescopersico) [inline ts/sagas/messages/loadMessage.ts:45]: Once we also migrate fetchMessage to v2 we can remove the converter.

File: ts/migration/fp-ts-converters.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import * as E from "fp-ts/lib/Either";
```
Comment (francescopersico) [inline ts/migration/fp-ts-converters.ts:1]: File with utility function to convert v1 <-> v2

File: ts/sagas/startup/watchLoadMessageWithRelationsSaga.ts [hunk 6]
```diff
@@ -25,11 +26,10 @@ export function* watchLoadMessageWithRelationsSaga(
...
   try {
-    const messageOrError: SagaCallReturnType<typeof loadMessage> = yield* call(
-      loadMessage,
-      getMessage,
-      messageId
-    );
+    const messageOrErrorv2: SagaCallReturnType<typeof loadMessage> =
+      yield* call(loadMessage, getMessage, messageId);
+
+    const messageOrError = eitherToV1(messageOrErrorv2);
```
Comment (francescopersico) [inline ts/sagas/startup/watchLoadMessageWithRelationsSaga.ts:32]: Use converter v2 -> v1

**Comment by CrisTofani** [comment]
closing since the merge of #4020

**Pull request closed without merging**